### PR TITLE
Implement AntiCheat System - AntiHacks:

### PR DESCRIPTION
**Pull request by kvipka**
-ASH - AntiSpeedHack (+anti teleport hack)
-AFH - AntiFlyHack
-DoubleJump
-Wallclimb
-FakeJumper
-FakeFlyingMode

This system has been tested on some big projects with big online (project names are hidden). 
**Issues:** No false positives

youtube videos : https://www.youtube.com/watch?v=aZP3qm9Ql0Y&list=PLUpIDd4JDWyOOJhJRcfaVL0RaRH1Ra5Xy

Note:
This system using only mathematic operations, without additional manager as AC2, no using additional memory.

Part of code can be shorter, when [PR](https://github.com/TrinityCore/TrinityCore/pull/18771) with ACK-packet handlers will complete.

ACK 1

### CODE REVIEW

File: sql_for_anticheat/2017_10_15_01_ASH_log.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+-- Strings for ASH + AFH allerts
+DELETE FROM `trinity_string` where `entry` IN (6617, 6618, 6619, 6620);
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES ('6617', 'AntiCheat: SpeedHack Detected for %s, normal distance for this time and speed = %f, distance from packet = %f');
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES ('6618', 'AntiCheat: FlyHack Detected for %s , player can not fly');
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES ('6619', 'AntiCheat: FlyHack Detected for %s, player has Swimming flag, but not in water');
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES ('6620', 'AntiCheat: FlyHack Detected for %s');
```
Comment (FrancescoBorzi) [inline sql_for_anticheat/2017_10_15_01_ASH_log.sql]: Please add the SQL files using the standard procedure: http://www.azerothcore.org/wiki/Dealing-with-SQL-files

Also can you merge those `INSERT` statements into a single one?

Comment (poszer) [reply sql_for_anticheat/2017_10_15_01_ASH_log.sql]:
```suggestion
INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES
('6617', 'AntiCheat: SpeedHack Detected for %s, normal distance for this time and speed = %f, distance from packet = %f'),
('6618', 'AntiCheat: FlyHack Detected for %s , player can not fly'),
('6619', 'AntiCheat: FlyHack Detected for %s, player has Swimming flag, but not in water'),
('6620', 'AntiCheat: FlyHack Detected for %s'); 
```

File: sql_for_anticheat/2018_04_05_00_AFH_fakeflying_mode_pinfo.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+﻿DELETE FROM trinity_string WHERE entry = 882;
```
Comment (poszer) [inline sql_for_anticheat/2018_04_05_00_AFH_fakeflying_mode_pinfo.sql]:
```suggestion
DELETE FROM `trinity_string` WHERE entry = 882;
```

File: src/server/game/Handlers/MovementHandler.cpp [hunk 3]
```diff
@@ -429,7 +454,65 @@ void WorldSession::HandleMovementOpcodes(WorldPacket & recvData)
...
+    }
+
+    if (sWorld->getBoolConfig(CONFIG_ANTICHEAT_FAKEJUMPER_ENABLED) && plrMover && !plrMover->UnderACKmount() && mover->IsFalling() && movementInfo.pos.GetPositionZ() > mover->GetPositionZ())
+    {
+        if (!plrMover->IsJumpingbyOpcode())
+        {
+            plrMover->SetJumpingbyOpcode(true);
+            plrMover->SetUnderACKmount();
+        }
+        else if (!movementInfo.flags & MOVEMENTFLAG_ONTRANSPORT)
```
Comment (FrancescoBorzi) [inline src/server/game/Handlers/MovementHandler.cpp]: Codacy tip: `Boolean result is used in bitwise operation. Clarify expression with parentheses.`

```cpp
else if (!(movementInfo.flags & MOVEMENTFLAG_ONTRANSPORT))
```

File: sql_for_anticheat/2018_03_25_00_AFH_logs.sql [hunk 4]
```diff
@@ -0,0 +1,4 @@
+-- Strings for AFH allerts
+DELETE FROM `trinity_string` where `entry` IN (6621, 6622);
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES ('6621', 'AntiCheat: DoubleJump Detected for %s');
```
Comment (poszer) [inline sql_for_anticheat/2018_03_25_00_AFH_logs.sql]:
```suggestion
INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES
('6621', 'AntiCheat: DoubleJump Detected for %s'),
('6622', 'AntiCheat: Fake Jumper Detected for %s');
```

**Comment by diluh** [comment]
@kvipka @FrancescoBorzi @poszer 

 I added this PR for testing [Warlock] - [Hunter] the pet disappears alone and also when descending from the mount

**Comment by disclosurez** [comment]
core won't build using this, lots of errors

**Comment by poszer** [comment]
> core won't build using this, lots of errors

Yeah, the PR is still in WIP

**Comment by lineagedr** [comment]
@kvipka I guess you're not selling it anymore? I've already replaced my old AC with this one and it's working like a charm detecting pretty much everything even desync. Thank you for the release kvipka.

**Comment by BarbzYHOOL** [comment]
@lineagedr can you help us fix the errors?

**Comment by wowmane** [comment]
@kvipka Thank You, Bro / good job!

Please will to done this commit.
I think very need to anti cheat for AC.

I use anticheat moudel but all can use the cheat hacks :laughing: 
http://www.azerothcore.org/modules-catalogue/details.html?id=76854840

**Comment by poszer** [comment]
> @kvipka Thank You, Bro / good job!
> 
> Tested and in the build, There are too many errors!
> 
> Please AC/DEVS Come For help...
> 
> Please will to done this commit.
> I think very good anti cheat for AC.

Feel free to fix issue

**Comment by milestorme** [comment]
Is this better than https://github.com/azerothcore/mod-anticheat . The module seems to work good

**Comment by poszer** [comment]
> Is this PR better than https://github.com/azerothcore/mod-anticheat . The module seems to work very good for me

Nah, module isn't good. Too much false reports. I'll not suggest you to use that modules. Use this one, this is real anticheat, and this is best one which exist atm (How i know) :D Thanks to @kvipka

**Comment by wowmane** [comment]
Yes @poszer , I use this moudel https://github.com/azerothcore/mod-anticheat, all can to use cheat hacks!

@milestorme give server address,i will to fast use cheat hacks ! fly/speed/teleport/jump

Thanks to @kvipka

**Comment by milestorme** [comment]
logon.wowtorn.net , have plenty of banned counts that tryed to use hacks , my players won’t hack because of the module and the fact that it seems to work , have a go if you like

**Comment by lineagedr** [comment]
> logon.wowtorn.net , have plenty of banned accounts that tryed to use hacks , my players won’t hack because of the module and the fact that it seems to work , have a go if you like

https://www.youtube.com/watch?v=ySJufN6Ggq4

**Comment by Winfidonarleyan** [comment]
@kvipka need add `override` to all virtual functions in class `Player`

File: src/server/game/Entities/Player/Player.h [hunk 5]
```diff
@@ -2534,8 +2561,8 @@ class Player : public Unit, public GridObject<Player>
         bool SetFeatherFall(bool apply, bool packetOnly = false);
         bool SetHover(bool enable, bool packetOnly = false);
 
-        bool CanFly() const { return m_movementInfo.HasMovementFlag(MOVEMENTFLAG_CAN_FLY); }
-
+        bool CanFly() const override { return m_movementInfo.HasMovementFlag(MOVEMENTFLAG_CAN_FLY); }
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Player/Player.h]: You add `override` to this function. Now you need to do so with all the virtual functions in this class

**Comment by wowmane** [comment]
> @kvipka I guess you're not selling it anymore? I've already replaced my old AC with this one and it's working like a charm detecting pretty much everything even desync. Thank you for the release kvipka.

@lineagedr
help us fix the errors?

**Comment by wowmane** [comment]
No one can solve the problem?  just need to fix warning errors  "build" !

Guys? 

This is good anticheat for azerothcore, please help to resolved.

**Comment by diluh** [comment]
This anticheat is not running 100% on AzerothCore

I have an active server and I put this anticheat, it is very good however there are still some fake checks that make the players take disconnect.

Some players reported that they sometimes kick, others reported me when they are dead can not walk and need to re-watch to be able to return to normal.

I was told that in the second boss of VOA players take kick

**Comment by wowmane** [comment]
@dilush

yes, but can to fixed.
set for boss spells 

if (((real_delta > 300.0f) && (real_delta > (allowed_delta * 3.0f))))

this is so good anticheat for azerothcore and will to imported commit.

**Comment by kvipka** [comment]
impressive...

if someone need a help or ask somnething about anticheat : only [this section](https://at-wow.org/forum/forums/anticheat-system-antihacks.81/)

**Comment by wowmane** [comment]
@kvipka 

for AZ build failed, please first of all fixed warning errors!

**Comment by poszer** [comment]
I'll update this tomorrow

**Comment by BarbzYHOOL** [comment]
hmmm :'(

**Pull request closed without merging**